psqlgraph/node.py

```diff
@@ -109,6 +109,52 @@ def get_edges(self):
         for edge_out in self.edges_out:
             yield edge_out
 
+    def bfs_children(self, edge_predicate=None, transform=None):
+        """
+        Perform a BFS, with `self` being the root node and perform arbitrary
+        node transformation along the way
+
+        :param edge_predicate: a predicate performed on an `edge` object in
+            order to decided whether to walk that edge or not
+        :type edge_predicate: func
+        :param transform: a transformation that needs to be done on the node
+        :type transform: func
+
+        :return: List[str]
+        """
+
+        if not callable(edge_predicate):
+            def edge_predicate(e):
+                return True
+
+        if not callable(transform):
+            def transform(node):
+                return node
+
+        traversal = []
+        visited = set()
+        queue = [self]
+
+        visited.add(self.node_id)
+
+        while queue:
+            current = queue.pop(0)
```

Comment:
`pop(0)` has a complexity of `O(n)` since it has to adjust the other members of the list, maybe use a double ended queue like `collections.deque`. Also shouldn't visited be populated after calling the transform function instead of L138 and L154?

Reply:
> pop(0) has a complexity of O(n) since it has to adjust the other members of the list, maybe use a double ended queue like `collections.deque`.

👍 

> Also shouldn't visited be populated after calling the transform function instead of L138 and L154?

reason is I wanted to avoid having same entries in the queue, which can happen if a node has multiple edges and it hasn't been processed yet.